draco/asp/hard.lp

```diff
@@ -1,24 +1,68 @@
 % ====== Hard constraints ======
 
-% @hard(text_mark_without_text_channel) Text mark requires text encoding.
-violation(text_mark_without_text_channel) :-
-    attribute((mark,type),M,text),
-    not attribute(mark_channel,M,text).
+% @hard(enc_type_valid) Primitive type has to support data type.
+violation(enc_type_valid) :-
+    attribute((field,type),F,(string;boolean)),
+    attribute((encoding,field),E,F),
+    attribute((encoding,type),E,(linear;log)).
 
-% @hard(text_channel_without_text_mark) Text channel requires text mark.
-violation(text_channel_without_text_mark) :-
-    attribute(mark_channel,M,text),
-    not attribute((mark,type),M,text).
+% @hard(log_non_positive) Cannot use log if the data is negative or zero.
+violation(log_non_positive) :-
+    attribute(extent,F,MIN,_),
+    attribute((encoding,field),E,F),
+    attribute((encoding,type),E,log),
+    MIN <= 0.
 
 % @hard(bin_and_aggregate) Cannot bin and aggregate.
 violation(bin_and_aggregate) :-
     attribute((encoding,binning),E,_),
     attribute((encoding,aggregate),E,_).
 
-% @hard(no_encodings) There has to be at least one encoding for every mark.
-violation(no_encodings) :-
-    entity(mark,_,M),
-    not entity(encoding,M,_).
+% @hard(aggregate_o_valid) Oridnal only supports min, max, and median.
```

Comment:
Typo.

Also, not what the actual constraint is. 

I would say categorical also does not really support min, max, and median.

Reply:
```suggestion
% @hard(aggregate_o_valid) Ordinal scale only supports min, max, and median.
```

Reply:
I did include categorical.

I also think this constraint is too indirect. Maybe this one should be about aggregation and data type, rather than the resulted scale?

Reply:
added two new constraints for aggregate and bin related to data types.

deleted `aggregate_o_valid` but we probably need a new version in the future.